federator/src/lib/FederatorERC.ts

```diff
@@ -0,0 +1,614 @@
+    transactionSender: TransactionSender;
+    fromBlock: number;
+    toBlock: number;
+    currentBlock: number;
+    mediumAndSmall: boolean;
+    confirmations: { mediumAmountConfirmations: number; largeAmountConfirmations: number };
+    sideChainConfig: ConfigChain;
+    federationFactory: FederationFactory;
+    allowTokensFactory: AllowTokensFactory;
+    bridgeFactory: BridgeFactory;
+  }) {
+    this.logger.trace(
+      `getLogsAndProcess started currentBlock: ${currentBlock}, fromBlock: ${fromBlock}, toBlock: ${toBlock}`,
+    );
+    if (fromBlock >= toBlock) {
+      this.logger.debug('getLogsAndProcess fromBlock >= toBlock', fromBlock, toBlock);
+      return;
+    }
+
+    const mainBridge = await bridgeFactory.getMainBridgeContract();
+
+    const recordsPerPage = 1000;
+    const numberOfPages = Math.ceil((toBlock - fromBlock) / recordsPerPage);
+    this.logger.debug(`Total pages ${numberOfPages}, blocks per page ${recordsPerPage}`);
+
+    let fromPageBlock = fromBlock;
+    for (let currentPage = 1; currentPage <= numberOfPages; currentPage++) {
+      let toPagedBlock = fromPageBlock + recordsPerPage - 1;
+      if (currentPage === numberOfPages) {
+        toPagedBlock = toBlock;
+      }
+      this.logger.debug(`Page ${currentPage} getting events from block ${fromPageBlock} to ${toPagedBlock}`);
+      const logs = await mainBridge.getPastEvents('Cross', {
+        fromBlock: fromPageBlock,
+        toBlock: toPagedBlock,
+      });
+      if (!logs) {
+        throw new Error('Failed to obtain the logs');
+      }
+
+      this.logger.info(`Found ${logs.length} logs`);
+      await this._processLogs({
+        sideChainId,
+        mainChainId,
+        transactionSender,
+        logs,
+        currentBlock,
+        mediumAndSmall,
+        confirmations,
+        sideChainConfig,
+        federationFactory,
+        allowTokensFactory,
+        bridgeFactory,
+      });
+      if (!mediumAndSmall) {
+        this._saveProgress(this.getLastBlockPath(sideChainId, mainChainId), toPagedBlock);
+      }
+      fromPageBlock = toPagedBlock + 1;
+    }
+  }
+
+  async checkFederatorIsMember(sideFedContract: IFederationV3 | IFederationV2, federatorAddress: string) {
+    const isMember = await typescriptUtils.retryNTimes(sideFedContract.isMember(federatorAddress));
+    if (!isMember) {
+      throw new Error(`This Federator addr:${federatorAddress} is not part of the federation`);
+    }
+  }
+
+  async processLog({
+    sideChainId,
+    mainChainId,
+    transactionSender,
+    log,
+    confirmations,
+    sideFedContract,
+    allowTokens,
+    currentBlock,
+    federatorAddress,
+    mediumAndSmall,
+    sideChainConfig,
+    bridgeFactory,
+  }: {
+    sideChainId: number;
+    mainChainId: number;
+    transactionSender: TransactionSender;
+    log: any;
+    confirmations: { largeAmountConfirmations: number; mediumAmountConfirmations: number };
+    sideFedContract: IFederationV3 | IFederationV2;
```

Comment:
Maybe it's easier to create a type called IFederation that is IFederationV3 | IFederationV2 and use that so the type is managed outside this file